### PR TITLE
Unroll transformation for TensorIterator

### PR DESCRIPTION
**Pull request by itikhono**

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
@itikhono am I right that these transformations are replacement for functions located in https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/legacy_api/include/net_pass.h ? Is it committed for 2021.1?

**Comment by itikhono** [comment]
> @itikhono am I right that these transformations are replacement for functions located in https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/legacy_api/include/net_pass.h ? Is it committed for 2021.1?

@ilya-lavrenov  Yes, you are right. Tickets: 31619, 31959

File: inference-engine/src/transformations/src/transformations/tensor_iterator_transformations/apply_transformations_to_ti_body.cpp [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+#include <memory>
+#include <vector>
+
+#include <ngraph/opsets/opset3.hpp>
+#include <ngraph/rt_info.hpp>
+#include <ngraph/graph_util.hpp>
+#include <ngraph/specialize_function.hpp>
+
+void ngraph::pass::ApplyTransformationsToTIBody::apply_transformations_to_ti_body(ngraph::pass::Manager& manager) {
+    auto tensor_iterator = std::make_shared<ngraph::pattern::op::Label>(ngraph::element::f32,
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/tensor_iterator_transformations/apply_transformations_to_ti_body.cpp:17]: Please use ngraph::pattern::wrap_type<ngraph::opset3::TensorIterator>();

File: inference-engine/tests/functional/inference_engine/ngraph_reader/ti.cpp [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 #include <string>
 #include "ngraph_reader_tests.hpp"
-TEST_F(NGraphReaderTests, ReadTensorIteratorNetwork) {
+TEST_F(NGraphReaderTests, DISABLED_ReadTensorIteratorNetwork) {
```
[the remark this answers is not in the document]
Comment (itikhono) [reply inference-engine/tests/functional/inference_engine/ngraph_reader/ti.cpp:7]: to check this PR with cpu plugin, these tests will be enabled in next commit

File: inference-engine/src/cldnn_engine/cldnn_engine.cpp [hunk 3]
```diff
@@ -102,15 +103,16 @@ InferenceEngine::ICNNNetwork::Ptr clDNNEngine::CloneAndTransformNetwork(const In
         manager.register_pass<ngraph::pass::ConvertOpSet3ToOpSet2>();
         manager.register_pass<ngraph::pass::ConvertOpSet2ToOpSet1>();
         manager.register_pass<ngraph::pass::ConvertOpSet1ToLegacy>();
-
-        manager.set_callback(transformations_callback);
-        manager.run_passes(nGraphFunc);
+        manager.register_pass<ngraph::pass::UnrollTensorIterator>();
```
Comment (GlebKazantaev) [inline inference-engine/src/cldnn_engine/cldnn_engine.cpp:106]: In this case you first apply all transformations to TI body and only after then you apply transformations to original function. Is it expected? Why we can not keep current order?

File: inference-engine/src/transformations/include/transformations/tensor_iterator_transformations/unroll_tensor_iterator.hpp [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+namespace ngraph {
+namespace pass {
+
+class TRANSFORMATIONS_API UnrollTensorIterator;
+
+}  // namespace pass
+}  // namespace ngraph
+
+/*
+ * Description:
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/transformations/tensor_iterator_transformations/unroll_tensor_iterator.hpp:23]: Please use doxygen description style and put some description :)

File: inference-engine/src/transformations/include/transformations/utils/utils.hpp [hunk 5]
```diff
@@ -44,6 +45,15 @@ bool has_op_with_type(const std::shared_ptr<const ngraph::Function> &function) {
     return false;
 }
 
+inline std::string create_ie_output_name(const ngraph::Output<ngraph::Node>& input) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/transformations/utils/utils.hpp:48]: `input` actually is an `output `

File: inference-engine/src/transformations/include/transformations/utils/utils.hpp [hunk 6]
```diff
@@ -44,6 +45,15 @@ bool has_op_with_type(const std::shared_ptr<const ngraph::Function> &function) {
...
+    const auto& prev_layer = input.get_node_shared_ptr();
+    auto out_idx = input.get_index();
+    std::string out_name = prev_layer->get_friendly_name();
+    if (prev_layer->get_output_size() != 1)
+        out_name += "." + std::to_string(out_idx);
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/transformations/utils/utils.hpp:53]:
```suggestion
        out_name += "." + std::to_string(output.get_index());
```

File: inference-engine/src/transformations/include/transformations/utils/utils.hpp [hunk 7]
```diff
@@ -44,6 +45,15 @@ bool has_op_with_type(const std::shared_ptr<const ngraph::Function> &function) {
     return false;
 }
 
+inline std::string create_ie_output_name(const ngraph::Output<ngraph::Node>& input) {
+    const auto& prev_layer = input.get_node_shared_ptr();
+    auto out_idx = input.get_index();
+    std::string out_name = prev_layer->get_friendly_name();
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/transformations/utils/utils.hpp:51]:
```suggestion
    std::string out_name = output.get_node()->get_friendly_name();
```
This must be faster

File: inference-engine/src/transformations/src/transformations/convert_opset3_to_opset2/convert_opset3_to_opset2.cpp [hunk 8]
```diff
@@ -16,6 +16,7 @@
 
 #include <ngraph/pass/manager.hpp>
 
+
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/convert_opset3_to_opset2/convert_opset3_to_opset2.cpp:19]: lets avoid this changes

File: inference-engine/src/transformations/src/transformations/tensor_iterator_transformations/unroll_tensor_iterator.cpp [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+#include <vector>
+
+#include <ngraph/graph_util.hpp>
+#include <ngraph/opsets/opset4.hpp>
+#include <ngraph/pattern/op/wrap_type.hpp>
+#include <ngraph/rt_info.hpp>
+
+void ngraph::pass::UnrollTensorIterator::unroll_tensor_iterator() {
+    auto tensor_iterator = ngraph::pattern::wrap_type<ngraph::opset4::TensorIterator>();
+    ngraph::graph_rewrite_callback callback = [this](pattern::Matcher& m) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/tensor_iterator_transformations/unroll_tensor_iterator.cpp:18]:
```suggestion
    ngraph::matcher_pass_callback callback = [this](pattern::Matcher& m) {
```

File: inference-engine/src/cldnn_engine/cldnn_engine.cpp (hunk 3 above)
Comment (itikhono) [reply inference-engine/src/cldnn_engine/cldnn_engine.cpp:106]: All transformations must be applied to the TensorIterator prior to calling UnrollTI due to performance reasons. Much faster to apply transformations to (1xbody) than to (n x body) after unrolling.

File: inference-engine/src/transformations/include/transformations/utils/utils.hpp (hunk 7 above)
Comment (itikhono) [reply inference-engine/src/transformations/include/transformations/utils/utils.hpp:50]: prev_layer is also used in the next line.

File: inference-engine/src/cldnn_engine/cldnn_engine.cpp (hunk 3 above)
Comment (GlebKazantaev) [reply inference-engine/src/cldnn_engine/cldnn_engine.cpp:106]: Yes I understand. I suggest to make this pipeline more clear:
```
 // Apply transformations to function with not unrolled TIs
 ngraph::pass::Manager manager;
 manager.register_pass<ngraph::pass::CommonOptimizations>();
 manager.register_pass<ngraph::pass::ConvertOpSet3ToOpSet2>();
 manager.register_pass<ngraph::pass::ConvertOpSet2ToOpSet1>();
 manager.register_pass<ngraph::pass::ConvertOpSet1ToLegacy>();
 manager.set_callback(transformations_callback);
 manager.run_passes(nGraphFunc);

 // Apply all transformations to TensorIterator body
 ngraph::pass::Manager ti_manager;
 ti_manager.register_pass<ngraph::pass::ApplyTransformationsToTIBody>(manager);
 // Unroll will be called after all conversions
 ti_manager.register_pass<ngraph::pass::UnrollTensorIterator>();
 ti_manager.run_passes(nGraphFunc);
```

File: inference-engine/src/cldnn_engine/cldnn_engine.cpp [hunk 10]
```diff
@@ -108,7 +109,10 @@ InferenceEngine::ICNNNetwork::Ptr clDNNEngine::CloneAndTransformNetwork(const In
 
         // Apply all transformations to TensorIterator body
         ngraph::pass::Manager ti_manager;
+        // Apply all transformations to TensorIterator body
```
Comment (GlebKazantaev) [inline inference-engine/src/cldnn_engine/cldnn_engine.cpp:112]: remove

File: inference-engine/src/transformations/include/transformations/tensor_iterator_transformations/unroll_tensor_iterator.hpp [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+ * the number of iterations of the TensorIterator layer, are created and connected to each other and to the external
+ * network. If the number of TensorIterator iterations is greater than 1, then additional Concat and Split layers
+ * are added to the network.
+ */
+
+
+class ngraph::pass::UnrollTensorIterator: public ngraph::pass::MatcherPass {
+public:
+    UnrollTensorIterator() : MatcherPass() {
+        unroll_tensor_iterator();
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/include/transformations/tensor_iterator_transformations/unroll_tensor_iterator.hpp:34]: put all code inside ctor

File: inference-engine/src/transformations/src/transformations/tensor_iterator_transformations/unroll_tensor_iterator.cpp [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+        const auto function = ti->get_body()->to_function();
+        auto num_iter = ti->get_num_iterations();
+
+        // negative value means inconsistent TI
+        if (num_iter <= -1) {
+            return false;
+        }
+
+        std::vector<std::shared_ptr<ngraph::Function>> body_functions(num_iter);
+        for (uint64_t idx = 0; idx < num_iter; ++idx) {
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/tensor_iterator_transformations/unroll_tensor_iterator.cpp:33]: Please add more comments

File: inference-engine/src/transformations/src/transformations/tensor_iterator_transformations/unroll_tensor_iterator.cpp [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+            body_functions[idx] = clone_function(*function);
+            for (auto &node : body_functions[idx]->get_ops()) {
+                node->set_friendly_name(ti->get_friendly_name() + "/" + std::to_string(idx + 1) + "/" + node->get_friendly_name());
+                copy_runtime_info(ti, node);
+            }
+        }
+
+        // Port map : inputs and back edges
+        for (const auto& desc : ti->get_input_descriptions()) {
+            std::string type_name = desc->get_type_info().name;
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/tensor_iterator_transformations/unroll_tensor_iterator.cpp:43]:
```suggestion
            const std::string & type_name = desc->get_type_info().name;
```

File: inference-engine/src/transformations/src/transformations/tensor_iterator_transformations/unroll_tensor_iterator.cpp [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+                    auto stride = output_desc->m_stride;
+                    for (uint64_t j = 0; j < num_iter; j++) {
+                        auto idx = stride > 0 ? j : num_iter - j - 1;
+                        std::shared_ptr<opset4::Result> result = body_functions[idx]->get_results()[output_desc->m_body_value_index];
+                        auto input_to_res = result->get_input_source_output(0);
+                        to_concat[j] = input_to_res;
+                    }
+                    auto concat = std::make_shared<ngraph::opset4::Concat>(to_concat, output_desc->m_axis);
+                    copy_runtime_info(ti, concat);
+                    concat->output(0).get_tensor().set_name(
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/tensor_iterator_transformations/unroll_tensor_iterator.cpp:134]: Please add comment with explanations

**Comment by gladilov-gleb** [comment]
@GlebKazantaev @itikhono @Maxim-Doronin @AndrewBakalinIntel @taka-no-me 

How it was merged without VPU team review? This request:

* Makes TI unrolling unconditional. VPU plugin does not unroll TI except for testing purposes.
* Calls TI unrolling twice: first during creation of executable network, then once again when nGraph is converted to CNN.
* Makes RNN cell combining impossible since TI would be unrolled before any attempt to combine will happen - it breaks performance as well.

CI didn't catch that because pre-commit does not run closed tests for myriad x (the only tests for now that validate TI logic in myriad - I guess it is so after pre-commit scope changes because closed tests do not contain *smoke* prefix).

I'll open PR with revert as a part of fix for #-37140

**Comment by itikhono** [comment]
> @GlebKazantaev @itikhono @Maxim-Doronin @AndrewBakalinIntel @taka-no-me
> 
> How it was merged without VPU team review? This request:
> 
> * Makes TI unrolling unconditional. VPU plugin does not unroll TI except for testing purposes.
> * Calls TI unrolling twice: first during creation of executable network, then once again when nGraph is converted to CNN.
> * Makes RNN cell combining impossible since TI would be unrolled before any attempt to combine will happen - it breaks performance as well.
> * While adds unrolling on nGraph side it does not remove unrolling call on CNN representation, so we actually trying to unroll TI not twice, but three times in different representations
> 
> CI didn't catch that because pre-commit does not run closed tests for myriad x (the only tests for now that validate TI logic in myriad - I guess it is so after pre-commit scope changes because closed tests do not contain _smoke_ prefix).
> 
> I'll open PR with revert as a part of fix for #-37140

I know about the issues above, it will be fixed at https://github.com/openvinotoolkit/openvino/pull/1594, CNN transformations will be removed. Sorry for the inconvenience.

File: inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp [hunk 15]
```diff
@@ -23,13 +23,13 @@
...
 
 #include <legacy/convert_function_to_cnn_network.hpp>
-#include <transformations/apply_transformations_to_ti_body.hpp>
 #include <transformations/common_optimizations/common_optimizations.hpp>
 #include <transformations/convert_opset1_to_legacy/convert_opset1_to_legacy.hpp>
 #include <transformations/convert_opset2_to_opset1/convert_opset2_to_opset1.hpp>
 #include <transformations/convert_opset3_to_opset2/convert_opset3_to_opset2.hpp>
 #include <transformations/convert_precision.hpp>
 #include <transformations/rt_info/fused_names_attribute.hpp>
+#include <transformations/tensor_iterator_transformations/apply_transformations_to_ti_body.hpp>
```
Comment (slyalin) [inline inference-engine/src/mkldnn_plugin/mkldnn_plugin.cpp:32]: Is it used in this file? I don't see other changes.

File: inference-engine/src/vpu/myriad_plugin/myriad_plugin.cpp [hunk 16]
```diff
@@ -52,7 +53,10 @@ ExecutableNetworkInternal::Ptr Engine::LoadExeNetworkImpl(
         manager.run_passes(function);
 
         ngraph::pass::Manager ti_manager;
+        // Apply all transformations to TensorIterator body
         ti_manager.register_pass<ngraph::pass::ApplyTransformationsToTIBody>(manager);
+        // Unroll should be called after all conversions
+        ti_manager.register_pass<ngraph::pass::UnrollTensorIterator>();
```
Comment (slyalin) [inline inference-engine/src/vpu/myriad_plugin/myriad_plugin.cpp:59]: Is it a new functionality for Myriad? I would expect that you remove something and insert this one

**Pull request merged**